client/media/CVideoHandler.cpp

```diff
@@ -0,0 +1,611 @@
+
+		sws_scale(sws, frame->data, frame->linesize, 0, getCodecContext()->height, data, linesize);
+		memcpy(surface->pixels, for_sws, pic_bytes);
+		av_free(for_sws);
+	}
+	return true;
+}
+
+bool CVideoInstance::videoEnded()
+{
+	return getCurrentFrame() == nullptr;
+}
+
+CVideoInstance::~CVideoInstance()
+{
+	sws_freeContext(sws);
+	SDL_DestroyTexture(textureYUV);
+	SDL_DestroyTexture(textureRGB);
+	SDL_FreeSurface(surface);
+}
+
+FFMpegStream::~FFMpegStream()
+{
+	av_frame_free(&frame);
+
+	avcodec_close(codecContext);
+	avcodec_free_context(&codecContext);
+
+	avcodec_close(codecContext);
+	avcodec_free_context(&codecContext);
+
+	avformat_close_input(&formatContext);
+	av_free(context);
+}
+
+Point CVideoInstance::size()
+{
+	if(!getCurrentFrame())
+		throw std::runtime_error("Invalid video frame!");
+
+	return Point(getCurrentFrame()->width, getCurrentFrame()->height);
+}
+
+void CVideoInstance::show(const Point & position, Canvas & canvas)
+{
+	if(sws == nullptr)
+		throw std::runtime_error("No video to show!");
+
+	CSDL_Ext::blitSurface(surface, canvas.getInternalSurface(), position);
+}
+
+double FFMpegStream::getCurrentFrameEndTime() const
+{
+#if(LIBAVUTIL_VERSION_MAJOR < 58)
+	auto packet_duration = frame->pkt_duration;
+#else
+	auto packet_duration = frame->duration;
+#endif
+	return (frame->pts + packet_duration) * av_q2d(formatContext->streams[streamIndex]->time_base);
+}
+
+double FFMpegStream::getCurrentFrameDuration() const
+{
+#if(LIBAVUTIL_VERSION_MAJOR < 58)
+	auto packet_duration = frame->pkt_duration;
+#else
+	auto packet_duration = frame->duration;
+#endif
+	return packet_duration * av_q2d(formatContext->streams[streamIndex]->time_base);
+}
+
+void CVideoInstance::tick(uint32_t msPassed)
+{
+	if(sws == nullptr)
+		throw std::runtime_error("No video to show!");
+
+	if(videoEnded())
+		throw std::runtime_error("Video already ended!");
+
+	frameTime += msPassed / 1000.0;
+
+	if(frameTime >= getCurrentFrameEndTime())
+		loadNextFrame();
+}
+
+struct FFMpegFormatDescription
+{
+	uint8_t sampleSizeBytes;
+	uint8_t wavFormatID;
+	bool isPlanar;
+};
+
+static FFMpegFormatDescription getAudioFormatProperties(int audioFormat)
+{
+	switch (audioFormat)
+	{
+		case AV_SAMPLE_FMT_U8:   return { 1, 1, false};
+		case AV_SAMPLE_FMT_U8P:  return { 1, 1, true};
+		case AV_SAMPLE_FMT_S16:  return { 2, 1, false};
+		case AV_SAMPLE_FMT_S16P: return { 2, 1, true};
+		case AV_SAMPLE_FMT_S32:  return { 4, 1, false};
+		case AV_SAMPLE_FMT_S32P: return { 4, 1, true};
+		case AV_SAMPLE_FMT_S64:  return { 8, 1, false};
+		case AV_SAMPLE_FMT_S64P: return { 8, 1, true};
+		case AV_SAMPLE_FMT_FLT:  return { 4, 3, false};
+		case AV_SAMPLE_FMT_FLTP: return { 4, 3, true};
+		case AV_SAMPLE_FMT_DBL:  return { 8, 3, false};
+		case AV_SAMPLE_FMT_DBLP: return { 8, 3, true};
+	}
+	throw std::runtime_error("Invalid audio format");
+}
+
+int FFMpegStream::findAudioStream() const
+{
+	for(int i = 0; i < formatContext->nb_streams; i++)
```

Comment:
As comment: As far as I know there is one special case. For the intro of the chronicles there are multiple audio streams. One per language. Even the english version contains the german audio stream for the intro.

Reply:
Noted, thanks for info.

Which Chronicle should I look up? Or any will do?

Reply:
Found it, 1st chronicle, Warlords of the Wasteland, Intro.bik and smk
5 different audio stream.

Unfortunately no information on language of each of these streams, will have to figure it out & hardcode it

Reply:
So in total we have:
HC1, HC3: english, french, german, italian, spanish
HC2, HC4, HC5, HC6: english
HC7, HC8: english, german

Reply:
Order should be:
```
english
french
german
italian
spanish
```

Used: https://translatedlabs.com/spoken-language-identifier

Reply:
Ah, you already figured it out. GitHub has not shown your last post to me. :(

Reply:
Still good to have confirmation from 2nd source :)

Didn't knew about such service so I transcribed English audio and then tried to compare auto-translated text to what I hear on other streams.

Reply:
I've checked it again. I think you mistaken HC3 and HC2?

HC1, HC2: english, french, german, italian, spanish
CD-Version (tested with german cd): HC3, HC4: english, french, german, italian, spanish
GOG-Version: HC3, HC4: english
HC5, HC6: english
HC7, HC8: english, german